src/lib/interrupt.c

```diff
@@ -33,11 +33,149 @@
 #include <sof/interrupt.h>
 #include <sof/interrupt-map.h>
 #include <sof/alloc.h>
+#include <sof/lock.h>
+#include <sof/list.h>
 #include <arch/interrupt.h>
 #include <platform/interrupt.h>
 #include <stdint.h>
 #include <stdlib.h>
 
+/*
+ * We have to invalidata cache of the list array, make sure no other data is
+ * lost while doing that
+ */
+static struct {
+	struct list_item list[PLATFORM_CORE_COUNT]
+		__attribute__((aligned(XCHAL_DCACHE_LINESIZE)));
+	spinlock_t lock __attribute__((aligned(XCHAL_DCACHE_LINESIZE)));
+} cascade_data;
+
+/*
+ * To avoid handling cache synchronisation manually when dealing with interrupts
+ * on multiple DSP cores, we enforce a local IRQ policy. Only cascading
+ * interrupt controller registration can be performed on an arbitrary core and
+ * will register the controller for all the cores. All other interrupt related
+ * functions must be called on the same core, where the interrupt is triggering.
+ * This isn't a problem for us, because all our tasks are anyway bound to their
+ * specific cores, no task migration is supported.
+ */
+
+/* register cascading IRQ controllers for all cores */
+int interrupt_cascade_register(const struct irq_cascade_tmpl *tmpl)
+{
+	struct irq_cascade_desc *cascade, *c;
+	int core = cpu_get_id();
+	struct list_item *list, *prev;
+	unsigned long flags;
+	unsigned int i, j;
+	int ret;
+
+	if (!tmpl->name)
+		return -EINVAL;
+
+	/* Make sure not to currupt adjacent data when invalidating cache */
+	cascade = (void *)((unsigned long)rmalloc(RZONE_SYS, SOF_MEM_CAPS_RAM,
```

Comment:
@lgirdwood @lyakh rmalloc_sys is automatically aligning to cache line size. I don't like this whole idea of master core allocating it's own system memory for other cores. We have separate system heaps for every core, so we don't need to preallocate anything and waste memory. Also I don't see, what this commit actually fixes.

Reply:
@tlauda it isn't a fix, it's a feature. Currently platforms have per-CPU static objects for on-SoC cascading interrupt controllers, see `static struct irq_desc dsp_irq[PLATFORM_CORE_COUNT][4]` in src/drivers/intel/cavs/interrupt.c. This patch series adds support for dynamic run-time interrupt controller registration. So, what you can do after this patch set is register such a controller from a task running on a specific core. But then other tasks, potentially running on other cores might need to request interrupts on that controller. So they somehow have to find it. The original approach was to use global descriptors for all cores, but that caused too much manual cache handling head ache. So we decided to keep the per-core interrupt descriptor approach. That's why during registration we now allocate those descriptor objects for each core. I didn't realise, that we now have per-core system memory heaps. It would be good to use them, although I think it isn't the unique case, that one of the cores (typically the boot core) does more work, especially during the initialisation, and does more allocations. So, maybe that heap for the boot core should be much larger. But if we do want to allocate those descriptors on specific core heaps, we can come up with several approaches too. E.g. (1) we could send IDCs to other cores to let them allocate descriptors too, this seems an overkill and over-engineering to me. (2) we could postpone allocating those descriptors until the first client interrupt is registered on a specific core. But that would require some global (for all CPUs) information about all available on the system cascading interrupt controllers. E.g. when one is registered, we could add a struct with all the non-volatile data to a global list, then when a client has to be registered on one of the cores, it could allocate a descriptor for the parent interrupt controller too. But then we would have a global list of those non-volatile data structs, that can be read and written from all cores, even if writing to it will happen very rarely. Maybe there are other ways too. @lgirdwood what would you prefer?

Reply:
@lyakh But even with your patch every core has its own interrupt descriptor, so it feels more natural to allocate it per core without master core preallocation. The code would be much simpler without all these cache operations and we wouldn't waste master core system heap.

Reply:
@lyakh can the IRQ core mapping be done at build time (via some Kconfig or platform header mapping of external IRQs)  ? if so this would simply the cache handling and also lessen the effort on using Zephyr/FreeRTOS as alternative RTOS.

Reply:
@lgirdwood oh... I don't know... If it could be done at build time, it would be extremely ugly and inflexible IMHO. Imagine a DW INTC (or a similar cascading interrupt controller), connected to one of levelX controllers and two devices on the INTC - one of them having to run on one and another one on another core. In principle one *could* fix this all at compile time, but I don't think that would be very pretty. Then again - we have pipeline configuration from the host over IPC, including core assignment, right? So, the host can decide at runtime which pipeline to run on which core and, respectively, which interrupts to use there.

Reply:
@lyakh so Zephyr already maps at build tine to save on footprint and complexity and we probably need to look at a similar trade off between cache management complexity and IRQ mapping flexibility. Internal local IRQs should be no problem, but the external IRQs can also come through other IRQ controller logic. So any mapping granularity would have to be at device level e.g. APL IRQ device level
```
/* IRQ Level 2 bits */
#define IRQ_BIT_LVL2_HP_GP_DMA0(x)	(x + 24)
#define IRQ_BIT_LVL2_WALL_CLK1		23
#define IRQ_BIT_LVL2_WALL_CLK0		22
#define IRQ_BIT_LVL2_L2_MEMERR		21
#define IRQ_BIT_LVL2_SHA256		16
#define IRQ_BIT_LVL2_L2_CACHE		15
#define IRQ_BIT_LVL2_IDC		8
#define IRQ_BIT_LVL2_HOST_IPC		7
#define IRQ_BIT_LVL2_CSME_IPC		6
#define IRQ_BIT_LVL2_PMC_IPC		5

/* IRQ Level 3 bits */
#define IRQ_BIT_LVL3_CODE_LOADER	31
#define IRQ_BIT_LVL3_HOST_STREAM_OUT(x)	(16 + x)
#define IRQ_BIT_LVL3_HOST_STREAM_IN(x)	(0 + x)

/* IRQ Level 4 bits */
#define IRQ_BIT_LVL4_LINK_STREAM_OUT(x)	(16 + x)
#define IRQ_BIT_LVL4_LINK_STREAM_IN(x)	(0 + x)

/* IRQ Level 5 bits */
#define IRQ_BIT_LVL5_LP_GP_DMA1(x)	(24 + x)
#define IRQ_BIT_LVL5_LP_GP_DMA0(x)	(16 + x)
#define IRQ_BIT_LVL5_DMIC(x)		6
#define IRQ_BIT_LVL5_SSP(x)		(0 + x)
```
This should be doable, and would only psuh the cache management into the CAVS IRQ controller logic ?

Reply:
@lyakh @tlauda in fact looking at the above again, it really only necessary to allow DMA and timer IRQs on the other cores as everything else can be done on core 0.

Reply:
@lgirdwood @lyakh This discussion just shows that there are no benefits of changing this part of code at all. If you want to remove static cascading interrupt controllers then do it, but allocate them dynamically on the core, that actually wants to use them and avoid this whole problem with cache. This commits has introduced the crazy amount of code, that previously was handled in one simple function.

Reply:
@lgirdwood @tlauda imagine this: you have a DMA controller or an I2C controller (serving a codec) or something else, attached to a 3rd or 4th level of cascading IRQ. You start a pipeline configuration process on the host to configure a pipeline on core 1. It goes on to register the IRQ for that DMA controller and finds, that the interrupt controller hasn't been instantiated yet. It calls the registration process for that controller and that again finds out, that its parent isn't registered on that core yet... This is the kind of recursion, that we already have, but at least now it doesn't perform the memory allocation - that is done earlier at the cascading controller registration time. That also allows to discard the non-volatile controller template data
```
struct irq_cascade_tmpl {
        const char *name;
        const struct irq_cascade_ops *ops;
        int irq;
        void (*handler)(void *arg);
};
```
If we switch over to allocating controller data objects at run-time on respective cores, we'll have to keep the above template data in a **global list**, synchronised for all cores. And as we have seen this doesn't work well on Xtensa. With the current approach we don't have global lists, we have just one global struct and per-core lists. With the proposed approach we would have a **global list** and I'd rather avoid that.

Reply:
@lyakh I guess this is our single thread now :)
 Can you condense the changes in features supported, init, memory allocation, list management and cache management is a small side by side comparison of current code vs this PR. This will make it easier to see the benefits etc.